homeassistant/components/xiaomi_aqara/config_flow.py

```diff
@@ -0,0 +1,188 @@
+                    return await self.async_step_settings()
+                if len(self.gateways) > 1:
+                    return await self.async_step_select()
+
+                errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=GATEWAY_CONFIG, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple aqara gateways found."""
+        errors = {}
+        if user_input is not None:
+            ip_adress = user_input["select_ip"]
+            self.selected_gateway = self.gateways[ip_adress]
+            return await self.async_step_settings()
+
+        select_scheme = vol.Schema(
+            {
+                vol.Required("select_ip"): vol.In(
+                    [gateway.ip_adress for gateway in self.gateways.values()]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_scheme, errors=errors
+        )
+
+    async def async_step_zeroconf(self, discovery_info):
+        """Handle zeroconf discovery."""
+        name = discovery_info.get("name")
+        self.host = discovery_info.get("host")
+        mac_address = discovery_info.get("properties", {}).get("mac")
+
+        if not name or not self.host or not mac_address:
+            return self.async_abort(reason="not_xiaomi_aqara")
+
+        # format mac (include semicolns and make uppercase)
+        mac_address = format_mac(mac_address)
+
+        # Check if the discovered device is an xiaomi aqara gateway.
+        if not name.startswith(ZEROCONF_GATEWAY):
+            _LOGGER.debug(
+                "Xiaomi device '%s' discovered with host %s, not identified as xiaomi aqara gateway",
+                name,
+                self.host,
+            )
+            return self.async_abort(reason="not_xiaomi_aqara")
+
+        unique_id = mac_address
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured({CONF_HOST: self.host})
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update({"title_placeholders": {"name": self.host}})
+
+        return await self.async_step_user()
+
+    async def async_step_settings(self, user_input=None):
+        """Specify settings and connect aqara gateway."""
+        errors = {}
+        if user_input is not None:
+            # get all required data
+            discovery_retry = user_input[CONF_DISCOVERY_RETRY]
+            name = user_input[CONF_NAME]
+            key = user_input.get(CONF_KEY)
+            ip_adress = self.selected_gateway.ip_adress
+            port = self.selected_gateway.port
+            sid = self.selected_gateway.sid
+            protocol = self.selected_gateway.proto
+
+            # get mac address
+            mac_address = await self.async_get_mac(ip_adress)
```

Comment:
Why can't we use the sid as unique id for the config entry? We don't want to use get_mac to look up mac address since that's not reliable.

Reply:
I think because the zeroconf discovery does not provide a sid and only a mac, but I will check this tonight.
Forgot the exact reason why I did it like this.....
If sid can be retrieved from zeroconf discovery I will use that

Reply:
Maybe we can query the gateway discovery in the zeroconf step and get the sid that way?

Reply:
I checked it and indeed the zeroconf discovery_info includes a mac_adress but does not include the sid....

Reply:
There is no direct way to get the sid from the gateway even if you already know the host/ip.
To get that you need to do a custom discovery of the gateway using the xiaomi_gateway library. Basically you need to sent UDP message `{"cmd":"whois"}` at multicast adresss '224.0.0.50' gateway_discovery_port 4321. Then the library wait for 5 seconds to gether all responses. Because you know the host already from the zeroconf discovery you can then match one of the gateways that responded to the zeroconf discovered gateway and then you know the required sid.

However that would include that 5 second discovery process inside the async_step_zeroconf, and I did not like that because if people do not use the xiaomi_aqara integration or have already configured it, I think it will go through that whole 5 second discovery process each time a zeroconf scan is done to then find out the SID is already known and ignored or configured. I think that might be a waste of resources.

That is why I opted to use the mac_adress for the unique_id because the mac is included in the zeroconf discovery_info, so no additional device comunication or discovery is required to then abort the config flow if the device is already set up or ignored using the unique_id based on mac.

Reply:
@MartinHjelmare If you like me to implement that 5 second discovery process inside the async_step_zeroconf to use sid as unique_id I will.
You can decide based on my previous comment....

Reply:
This is the zeroconf discovery_info:
`{'host': '192.168.1.IP', 'port': 54321, 'hostname': 'lumi-gateway-v3_miio12345678.local.', 'type': '_miio._udp.local.', 'name': 'lumi-gateway-v3_miio12345678._miio._udp.local.', 'properties': {'_raw': {'epoch': b'1', 'mac': b'abcdefghifjk'}, 'epoch': '1', 'mac': 'abcdefghifjk'}}`

Reply:
Since we can't get a unique_id easily, I think we should remove the zeroconf step here and let the default zeroconf step handle the zeroconf discovery. That will allow zeroconf discovery until a config entry is created for this domain.

We can then change to use the `sid` as unique_id for the user configured config entries.

Reply:
The default zeroconf step will not work since that will detect lots of other xiaomi products that are not Xiaomi Gateways and can therefore not be used with this integration.
Other products that will be discovered on the same zeroconf service but be filtered out:
- xiaomi air purifiers
- xioami air humidifiers
- xiaomi alarm clock
- xiaomi camera's
- xiaomi speakers
- xiaomi robot vacuum cleaners
- xioami wifi plugs/switches/sockets
and lots more....

This integration only handels xiaomi gateways and there connected zigbee devices.

Reply:
@MartinHjelmare I went though everything again and I discovered that I am stupid.
It turns out the sid and the mac are the same thing....
Xiaomi uses sid as device identification, but for the gateway they apperently use the mac_adress as identification. so for a gateway the sid and mac are the same.

Other devices such as the connected zigbee devices do not have an mac, but do have a sid which also has a larger number of characters than a mac.

Anyway I will update the code since sid and mac can be used interchangable for the gateway....

Reply:
I removed the getmac depandance